src/Microsoft.OData.Edm/Vocabularies/Annotations/EdmDirectValueAnnotationsManager.cs

```diff
@@ -197,129 +201,77 @@ private static void SetAnnotation(IEnumerable<IEdmDirectValueAnnotation> immutab
 
             if (transientAnnotations == null)
             {
-                transientAnnotations = newAnnotation;
-                return;
-            }
-
-            IEdmDirectValueAnnotation singleAnnotation = transientAnnotations as IEdmDirectValueAnnotation;
-            if (singleAnnotation != null)
-            {
-                if (singleAnnotation.NamespaceUri == namespaceName && singleAnnotation.Name == localName)
-                {
-                    transientAnnotations = newAnnotation;
-                }
-                else
-                {
-                    transientAnnotations = VersioningList<IEdmDirectValueAnnotation>.Create().Add(singleAnnotation).Add(newAnnotation);
-                }
-
+                transientAnnotations = VersioningList<IEdmDirectValueAnnotation>.Create().Add(newAnnotation);
                 return;
             }
 
-            VersioningList<IEdmDirectValueAnnotation> annotationsList = (VersioningList<IEdmDirectValueAnnotation>)transientAnnotations;
-            for (int index = 0; index < annotationsList.Count; index++)
+            for (int index = 0; index < transientAnnotations.Count; index++)
             {
-                IEdmDirectValueAnnotation existingAnnotation = annotationsList[index];
+                IEdmDirectValueAnnotation existingAnnotation = transientAnnotations[index];
                 if (existingAnnotation.NamespaceUri == namespaceName && existingAnnotation.Name == localName)
                 {
-                    annotationsList = annotationsList.RemoveAt(index);
+                    transientAnnotations = transientAnnotations.RemoveAt(index);
                     break;
                 }
             }
 
-            transientAnnotations = annotationsList.Add(newAnnotation);
+            transientAnnotations = transientAnnotations.Add(newAnnotation);
         }
 
-        private static IEdmDirectValueAnnotation FindTransientAnnotation(object transientAnnotations, string namespaceName, string localName)
+        private static IEdmDirectValueAnnotation FindTransientAnnotation(VersioningList<IEdmDirectValueAnnotation> transientAnnotations, string namespaceName, string localName)
         {
             if (transientAnnotations != null)
             {
-                IEdmDirectValueAnnotation singleAnnotation = transientAnnotations as IEdmDirectValueAnnotation;
-                if (singleAnnotation != null)
+                // This method runs very hot in AGS, it is important to optimize it as much as possible.

```

Comment:
>This method runs very hot in AGS [](http://example.com/codeflow?start=19&length=32)

We should remove an product-specific references from comments.  Maybe replace with something like "This method runs very hot in schemas where direct value annotations are used."